File: 356_densenet_trading/rust_densenet_trading/src/data/bybit_client.rs
//! Bybit API Client
//!
//! Client for fetching cryptocurrency market data from Bybit exchange.

use anyhow::{anyhow, Result};
use chrono::{DateTime, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::Duration;

use super::candle::Candle;
use super::orderbook::{OrderBook, OrderBookLevel};

/// Bybit API client for fetching market data
#[derive(Debug, Clone)]
pub struct BybitClient {
    /// HTTP client
    client: Client,

    /// Base URL for API
    base_url: String,

    /// Request timeout
    timeout: Duration,
}

/// Bybit API response wrapper
#[derive(Debug, Deserialize)]
struct BybitResponse<T> {
    #[serde(rename = "retCode")]
    ret_code: i32,
    #[serde(rename = "retMsg")]
    ret_msg: String,
    result: T,
}

/// Kline (candlestick) response
#[derive(Debug, Deserialize)]
struct KlineResult {
    category: String,
    symbol: String,
    list: Vec<Vec<String>>,
}

/// Order book response
#[derive(Debug, Deserialize)]
struct OrderBookResult {
    s: String, // symbol
    b: Vec<Vec<String>>, // bids
    a: Vec<Vec<String>>, // asks
    ts: u64, // timestamp
    u: u64, // update id
}

impl BybitClient {
    /// Create a new Bybit client
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .expect("Failed to create HTTP client"),
            base_url: "https://api.bybit.com".to_string(),
            timeout: Duration::from_secs(30),
        }
    }

    /// Create a client for testnet
    pub fn testnet() -> Self {
        Self {
            client: Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .expect("Failed to create HTTP client"),
            base_url: "https://api-testnet.bybit.com".to_string(),
            timeout: Duration::from_secs(30),
        }
    }

    /// Fetch kline (candlestick) data
    ///
    /// # Arguments
    /// * `symbol` - Trading pair (e.g., "BTCUSDT")
    /// * `interval` - Candle interval (1, 3, 5, 15, 30, 60, 120, 240, 360, 720, D, W, M)
    /// * `limit` - Number of candles (max 1000)
    /// * `start` - Start time (optional)
    /// * `end` - End time (optional)
    pub async fn get_klines(
        &self,
        symbol: &str,
        interval: &str,
        limit: usize,
        start: Option<u64>,
        end: Option<u64>,
    ) -> Result<Vec<Candle>> {
        let mut url = format!(
            "{}/v5/market/kline?category=linear&symbol={}&interval={}&limit={}",
            self.base_url,
            symbol,
            interval,
            limit.min(1000)
        );

        if let Some(s) = start {
            url.push_str(&format!("&start={}", s));
        }
        if let Some(e) = end {
            url.push_str(&format!("&end={}", e));
        }

        let response: BybitResponse<KlineResult> = self
            .client
            .get(&url)
            .send()
            .await?
            .json()
            .await?;

        if response.ret_code != 0 {
            return Err(anyhow!("Bybit API error: {}", response.ret_msg));
        }

        let candles: Vec<Candle> = response
            .result
            .list
            .into_iter()
            .filter_map(|row| {
                if row.len() >= 7 {
                    Some(Candle {
                        timestamp: row[0].parse().unwrap_or(0),
                        open: row[1].parse().unwrap_or(0.0),
                        high: row[2].parse().unwrap_or(0.0),
                        low: row[3].parse().unwrap_or(0.0),
                        close: row[4].parse().unwrap_or(0.0),
                        volume: row[5].parse().unwrap_or(0.0),
                        turnover: row[6].parse().unwrap_or(0.0),
                    })
                } else {
                    None
                }
            })
            .collect();

        Ok(candles)
    }

    /// Fetch order book
    ///
    /// # Arguments
    /// * `symbol` - Trading pair (e.g., "BTCUSDT")
    /// * `limit` - Depth limit (1, 25, 50, 100, 200)
    pub async fn get_orderbook(&self, symbol: &str, limit: usize) -> Result<OrderBook> {
        let url = format!(
            "{}/v5/market/orderbook?category=linear&symbol={}&limit={}",
            self.base_url,
            symbol,
            limit.min(200)
        );

        let response: BybitResponse<OrderBookResult> = self
            .client
            .get(&url)
            .send()
            .await?
            .json()
            .await?;

        if response.ret_code != 0 {
            return Err(anyhow!("Bybit API error: {}", response.ret_msg));
        }

        let result = response.result;

        let bids: Vec<OrderBookLevel> = result
            .b
            .into_iter()
            .filter_map(|row| {
                if row.len() >= 2 {
                    Some(OrderBookLevel {
                        price: row[0].parse().unwrap_or(0.0),
                        quantity: row[1].parse().unwrap_or(0.0),
                    })
                } else {
                    None
                }
            })
            .collect();

        let asks: Vec<OrderBookLevel> = result
            .a
            .into_iter()
            .filter_map(|row| {
                if row.len() >= 2 {
                    Some(OrderBookLevel {
                        price: row[0].parse().unwrap_or(0.0),
                        quantity: row[1].parse().unwrap_or(0.0),
                    })
                } else {
                    None
                }
            })
            .collect();

        Ok(OrderBook::new(result.s, bids, asks, result.ts))
    }

    /// Fetch ticker information
    pub async fn get_ticker(&self, symbol: &str) -> Result<TickerInfo> {
        let url = format!(
            "{}/v5/market/tickers?category=linear&symbol={}",
            self.base_url, symbol
        );

        let response: serde_json::Value = self
            .client
            .get(&url)
            .send()
            .await?
            .json()
            .await?;

        let ret_code = response["retCode"].as_i64().unwrap_or(-1);
        if ret_code != 0 {
            return Err(anyhow!("Bybit API error: {:?}", response["retMsg"]));
        }

        let list = &response["result"]["list"];
        if let Some(ticker) = list.get(0) {
            Ok(TickerInfo {
                symbol: ticker["symbol"].as_str().unwrap_or("").to_string(),
                last_price: ticker["lastPrice"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                bid_price: ticker["bid1Price"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                ask_price: ticker["ask1Price"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                volume_24h: ticker["volume24h"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                turnover_24h: ticker["turnover24h"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                high_24h: ticker["highPrice24h"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                low_24h: ticker["lowPrice24h"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                price_change_24h: ticker["price24hPcnt"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                funding_rate: ticker["fundingRate"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
                open_interest: ticker["openInterest"].as_str().unwrap_or("0").parse().unwrap_or(0.0),
            })
        } else {
            Err(anyhow!("No ticker data found"))
        }
    }

    /// Fetch historical funding rate
    pub async fn get_funding_rate_history(
        &self,
        symbol: &str,
        limit: usize,
    ) -> Result<Vec<FundingRate>> {
        let url = format!(
            "{}/v5/market/funding/history?category=linear&symbol={}&limit={}",
            self.base_url,
            symbol,
            limit.min(200)
        );

        let response: serde_json::Value = self
            .client
            .get(&url)
            .send()
            .await?
            .json()
            .await?;

        let ret_code = response["retCode"].as_i64().unwrap_or(-1);
        if ret_code != 0 {
            return Err(anyhow!("Bybit API error: {:?}", response["retMsg"]));
        }

        let list = response["result"]["list"].as_array();
        let funding_rates: Vec<FundingRate> = list
            .map(|arr| {
                arr.iter()
                    .filter_map(|item| {
                        Some(FundingRate {
                            symbol: item["symbol"].as_str()?.to_string(),
                            funding_rate: item["fundingRate"].as_str()?.parse().ok()?,
                            funding_rate_timestamp: item["fundingRateTimestamp"]
                                .as_str()?
                                .parse()
                                .ok()?,
                        })
                    })
                    .collect()
            })
            .unwrap_or_default();

        Ok(funding_rates)
    }
}

impl Default for BybitClient {
    fn default() -> Self {
        Self::new()
    }
}

/// Ticker information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TickerInfo {
    pub symbol: String,
    pub last_price: f64,
    pub bid_price: f64,
    pub ask_price: f64,
    pub volume_24h: f64,
    pub turnover_24h: f64,
    pub high_24h: f64,
    pub low_24h: f64,
    pub price_change_24h: f64,
    pub funding_rate: f64,
    pub open_interest: f64,
}

/// Funding rate data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FundingRate {
    pub symbol: String,
    pub funding_rate: f64,
    pub funding_rate_timestamp: u64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_client_creation() {
        let client = BybitClient::new();
        assert!(client.base_url.contains("bybit.com"));
    }

    #[test]
    fn test_testnet_client() {
        let client = BybitClient::testnet();
        assert!(client.base_url.contains("testnet"));
    }
}

File: 69_llm_earnings_call_analysis/python/bybit_client.py
"""
Bybit API Client for Cryptocurrency Market Data

This module provides a client for fetching market data from Bybit exchange,
which can be used for analyzing crypto project announcements similar to
earnings calls.
"""

import requests
from typing import List, Dict, Optional
from datetime import datetime
from dataclasses import dataclass


@dataclass
class Candle:
    """OHLCV candle data"""
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: float


@dataclass
class Ticker:
    """Current ticker information"""
    symbol: str
    last_price: float
    bid: float
    ask: float
    volume_24h: float
    price_change_24h: float


class BybitClient:
    """
    Client for fetching market data from Bybit exchange

    Supports fetching candlestick data, ticker information, and
    other market data useful for crypto trading analysis.
    """

    BASE_URL = "https://api.bybit.com"

    def __init__(self, testnet: bool = False):
        """
        Initialize Bybit client

        Args:
            testnet: If True, use testnet instead of mainnet
        """
        if testnet:
            self.base_url = "https://api-testnet.bybit.com"
        else:
            self.base_url = self.BASE_URL

        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json'
        })

    def get_klines(self,
                   symbol: str,
                   interval: str,
                   limit: int = 200,
                   start_time: Optional[int] = None,
                   end_time: Optional[int] = None) -> List[Candle]:
        """
        Get candlestick data from Bybit

        Args:
            symbol: Trading pair (e.g., "BTCUSDT")
            interval: Candlestick interval:
                - "1", "3", "5", "15", "30" (minutes)
                - "60", "120", "240", "360", "720" (minutes)
                - "D" (daily), "W" (weekly), "M" (monthly)
            limit: Number of candles to fetch (max 1000)
            start_time: Start timestamp in milliseconds
            end_time: End timestamp in milliseconds

        Returns:
            List of Candle objects in chronological order
        """
        endpoint = f"{self.base_url}/v5/market/kline"

        params = {
            "category": "linear",
            "symbol": symbol,
            "interval": interval,
            "limit": min(limit, 1000)
        }

        if start_time:
            params["start"] = start_time
        if end_time:
            params["end"] = end_time

        response = self.session.get(endpoint, params=params)
        data = response.json()

        if data['retCode'] != 0:
            raise Exception(f"Bybit API error: {data['retMsg']}")

        candles = []
        for item in data['result']['list']:
            candles.append(Candle(
                timestamp=datetime.fromtimestamp(int(item[0]) / 1000),
                open=float(item[1]),
                high=float(item[2]),
                low=float(item[3]),
                close=float(item[4]),
                volume=float(item[5])
            ))

        return candles[::-1]  # Reverse to chronological order

    def get_ticker(self, symbol: str) -> Ticker:
        """
        Get current ticker information

        Args:
            symbol: Trading pair (e.g., "BTCUSDT")

        Returns:
            Ticker object with current market data
        """
        endpoint = f"{self.base_url}/v5/market/tickers"

        params = {
            "category": "linear",
            "symbol": symbol
        }

        response = self.session.get(endpoint, params=params)
        data = response.json()

        if data['retCode'] != 0:
            raise Exception(f"Bybit API error: {data['retMsg']}")

        ticker = data['result']['list'][0]
        return Ticker(
            symbol=ticker['symbol'],
            last_price=float(ticker['lastPrice']),
            bid=float(ticker['bid1Price']),
            ask=float(ticker['ask1Price']),
            volume_24h=float(ticker['volume24h']),
            price_change_24h=float(ticker['price24hPcnt'])
        )

    def get_orderbook(self, symbol: str, limit: int = 25) -> Dict:
        """
        Get order book data

        Args:
            symbol: Trading pair
            limit: Depth of orderbook (max 500)

        Returns:
            Dictionary with bids and asks
        """
        endpoint = f"{self.base_url}/v5/market/orderbook"

        params = {
            "category": "linear",
            "symbol": symbol,
            "limit": min(limit, 500)
        }

        response = self.session.get(endpoint, params=params)
        data = response.json()

        if data['retCode'] != 0:
            raise Exception(f"Bybit API error: {data['retMsg']}")

        result = data['result']
        return {
            'symbol': result['s'],
            'timestamp': datetime.fromtimestamp(int(result['ts']) / 1000),
            'bids': [(float(b[0]), float(b[1])) for b in result['b']],
            'asks': [(float(a[0]), float(a[1])) for a in result['a']]
        }

    def get_recent_trades(self, symbol: str, limit: int = 50) -> List[Dict]:
        """
        Get recent trades

        Args:
            symbol: Trading pair
            limit: Number of trades (max 1000)

        Returns:
            List of recent trades
        """
        endpoint = f"{self.base_url}/v5/market/recent-trade"

        params = {
            "category": "linear",
            "symbol": symbol,
            "limit": min(limit, 1000)
        }

        response = self.session.get(endpoint, params=params)
        data = response.json()

        if data['retCode'] != 0:
            raise Exception(f"Bybit API error: {data['retMsg']}")

        trades = []
        for trade in data['result']['list']:
            trades.append({
                'id': trade['execId'],
                'price': float(trade['price']),
                'size': float(trade['size']),
                'side': trade['side'],
                'timestamp': datetime.fromtimestamp(int(trade['time']) / 1000)
            })

        return trades

    def get_funding_rate(self, symbol: str) -> Dict:
        """
        Get funding rate for perpetual contracts

        Args:
            symbol: Trading pair

        Returns:
            Dictionary with funding rate information
        """
        endpoint = f"{self.base_url}/v5/market/tickers"

        params = {
            "category": "linear",
            "symbol": symbol
        }

        response = self.session.get(endpoint, params=params)
        data = response.json()

        if data['retCode'] != 0:
            raise Exception(f"Bybit API error: {data['retMsg']}")

        ticker = data['result']['list'][0]
        return {
            'symbol': symbol,
            'funding_rate': float(ticker.get('fundingRate', 0)),
            'next_funding_time': datetime.fromtimestamp(
                int(ticker.get('nextFundingTime', 0)) / 1000
            ) if ticker.get('nextFundingTime') else None
        }

    def calculate_returns(self, candles: List[Candle]) -> List[float]:
        """
        Calculate returns from candle data

        Args:
            candles: List of Candle objects

        Returns:
            List of returns (percentage change)
        """
        returns = []
        for i in range(1, len(candles)):
            ret = (candles[i].close - candles[i-1].close) / candles[i-1].close
            returns.append(ret)
        return returns

    def calculate_volatility(self, candles: List[Candle], window: int = 20) -> float:
        """
        Calculate historical volatility

        Args:
            candles: List of Candle objects
            window: Rolling window size

        Returns:
            Annualized volatility
        """
        import numpy as np

        returns = self.calculate_returns(candles)

        if len(returns) < window:
            return 0.0

        # Use most recent window
        recent_returns = returns[-window:]
        daily_vol = np.std(recent_returns)

        # Annualize (assuming crypto trades 365 days)
        return daily_vol * np.sqrt(365)


def fetch_crypto_data_for_analysis(symbol: str = "BTCUSDT",
                                   interval: str = "D",
                                   days: int = 30) -> Dict:
    """
    Fetch cryptocurrency data for analysis

    This function fetches market data that can be used alongside
    crypto project announcement analysis (similar to earnings calls).

    Args:
        symbol: Trading pair
        interval: Candle interval
        days: Number of days of history

    Returns:
        Dictionary with market data and metrics
    """
    client = BybitClient()

    # Fetch candles
    candles = client.get_klines(symbol, interval, limit=days)

    # Get current ticker
    ticker = client.get_ticker(symbol)

    # Calculate metrics
    returns = client.calculate_returns(candles)
    volatility = client.calculate_volatility(candles)

    # Calculate simple statistics
    import numpy as np

    return {
        'symbol': symbol,
        'current_price': ticker.last_price,
        'price_change_24h': ticker.price_change_24h,
        'volume_24h': ticker.volume_24h,
        'candles': candles,
        'returns': returns,
        'volatility': volatility,
        'avg_return': np.mean(returns) if returns else 0,
        'max_return': max(returns) if returns else 0,
        'min_return': min(returns) if returns else 0,
        'positive_days': sum(1 for r in returns if r > 0),
        'negative_days': sum(1 for r in returns if r < 0)
    }


if __name__ == "__main__":
    # Example usage
    print("Fetching Bybit data...")

    try:
        data = fetch_crypto_data_for_analysis("BTCUSDT", "D", 30)

        print(f"\n=== {data['symbol']} Market Data ===")
        print(f"Current Price: ${data['current_price']:,.2f}")
        print(f"24h Change: {data['price_change_24h']*100:.2f}%")
        print(f"24h Volume: ${data['volume_24h']:,.0f}")
        print(f"\n=== Statistics (Last 30 Days) ===")
        print(f"Volatility (Annualized): {data['volatility']*100:.1f}%")
        print(f"Average Daily Return: {data['avg_return']*100:.3f}%")
        print(f"Best Day: {data['max_return']*100:.2f}%")
        print(f"Worst Day: {data['min_return']*100:.2f}%")
        print(f"Positive Days: {data['positive_days']}")
        print(f"Negative Days: {data['negative_days']}")

    except Exception as e:
        print(f"Error fetching data: {e}")
        print("\nNote: This example requires internet connection to Bybit API.")
